OSDOCS 12363: Added migration from v1 to v2 section in oc-mirror docs (#87754)
openshift/openshift-docs: modules/oc-mirror-migration-process.adoc

// Module included in the following assemblies:
//
// * oc-mirror-migration-v1-to-v2.adoc

:_mod-docs-content-type: PROCEDURE
[id="oc-mirror-migration-process_{context}"]
= Migrating from oc-mirror plugin v1 to oc-mirror plugin v2

To migrate from oc-mirror plugin v1 to v2, manually update the `ImageSetConfiguration`, modify mirroring commands, and clean up v1 artifacts. Follow these steps to complete the migration.

> . Modify the API version and remove deprecated fields in your `ImageSetConfiguration`. RSee the examples below for the configuration before and after the update:

Reviewer: Typo here for `RSee`

mirror:
  platform:
    channels:
      - name: stable-4.17
    graph: true
  helm:
    repositories:
      - name: sbo
        url: https://redhat-developer.github.io/service-binding-operator-helm-chart/
  additionalImages:
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/openshifttest/hello-openshift@sha256:example_hash
  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
storageConfig:  # REMOVE this field in v2
  local:
    path: /var/lib/oc-mirror
----
+
.Example output with oc-mirror plugin v2 configuration:
[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
      - name: stable-4.17
    graph: true
  helm:
    repositories:
      - name: sbo
        url: https://redhat-developer.github.io/service-binding-operator-helm-chart/
  additionalImages:
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/openshifttest/hello-openshift@sha256:example_hash
  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
----

> . In v2, mirrored content is stored in a different location than v1. You must check the cluster-resources directory inside their working directory for IDMS, ITMS, `CatalogSource`, and `ClusterCatalog` using the following command:

Reviewer 2: s/thier/the

+
[source,terminal]
----
$ ls <v2_workspace>/working-dir/cluster-resources/
----

> . To remove images mirrored using oc-mirror plugin v1, follow these steps:

Reviewer 3: ```suggestion
. [OPTIONAL] To remove images mirrored using oc-mirror plugin v1, follow these steps:
```

Author: Done

+
[IMPORTANT]
====
oc-mirror plugin v2 does not automatically prune the destination registry, unlike oc-mirror plugin v1. If you want to clean up images no longer needed, use the delete functionality in v2 with `--delete-v1-images`.

Once all images mirrored with oc-mirror plugin v1 are removed, you no longer need to use this flag. If you need to delete images mirrored with oc-mirror plugin v2, do not set `--delete-v1-images`.
====

.. Mirror the image using oc-mirror plugin v1.

.. Upgrade the `ImageSetConfiguration` API version from `v1alpha2` (v1) to `v2alpha1` (v2), then run the following command:

> +

Reviewer 2: should we also say if `storageConfig` is present remove that as well ?

Author: @sherine-k  @aguidirh @lmzuccarelli thoughts on this?

Reviewer 3: I am not sure if it is needed as @subhtk already added in the `Configuration changes` sub-section the following sentence:

```
storageConfig must be removed in oc-mirror plugin v2.
```

Reviewer 2: I think better to add as customers would be following the listed steps to do the migration.

Reviewer 3: So a little reminder saying that `storageConfig` is not a valid field in the DeleteImageSetConfiguration would work as well.

Author: Done

[source,terminal]
----
$ oc-mirror -c isc.yaml file://some-dir --v2
----

.. Generate a delete manifest and delete v1 images using the following command:
+

> [source,terminal]

Reviewer 2: Do you think we need to say `update kind in imageSetConfig as DeleteImageSetConfiguration` and change `mirror` to `delete` before starting off with the delete process ?

Author: cc: @sherine-k @aguidirh @lmzuccarelli

Reviewer 3: I would add a link on how to generate a delete image set configuration, in this way we keep the migration section shrink.

Reviewer 2: @aguidirh that works !!

Author: Since I have already added the reference for delete procedure. I'll mark this as resolved.

Reviewer 2: @lmzuccarelli i agree to have an example, if we all agree and just having that example does not cause any harm let us add it instead of juggling between the sections may be ?

----
$ oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://registry.example:5000  --v2
----
+
For more information about deleting images, see "Deletion of images from your disconnected environment".

.. Delete images based on the generated manifest:
+
[source,terminal]
----
$ oc-mirror delete --delete-yaml-file some-dir/working-dir/delete/delete-images.yaml docker://registry.example:5000 --v2
----

> .Verification

Reviewer 2: Going through the flow this verification looks like a disconnection to me, can we add it before the delete ?  WDYT ?
cc: @aguidirh

Reviewer 3: The `Migrating from oc-mirror plugin v1 to oc-mirror plugin v2` section is composed of 3 steps and the third one is optional. 

I agree with you that the verification section could come after the second step (since the third one is optional).

Is there a way to put the `Verification` after the second step @subhtk without making the steps confusing?

Author: @seb your insights?

Reviewer 4: I can see multiple sides to this matter. Technically speaking, verification sections always have to appear at the end of the procedure, and so if these things need to be verified before step 3, you would need to move step 3 to its own procedure.

However, I also see that the nature of this procedure involves checking things (step 2 is just checking that resources exist in a directory), so I also wouldn't see anything super wrong with moving the verification steps into this procedure and simply not having a verification section of this procedure.

Author: Moved the verification steps before the optional step for clear, concise procedure.

  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
----

. In oc-mirror plugin v2, mirrored content is stored in a different location than v1. You must check the cluster-resources directory inside the working directory for IDMS, ITMS, `CatalogSource`, and `ClusterCatalog` using the following command:
+
[source,terminal]
----
$ ls <v2_workspace>/working-dir/cluster-resources/
----

. Once the migration is complete, verify that mirrored images and catalogs are available. Make sure that:
- No errors or warnings occurred during mirroring.
- No error file was generated (`working-dir/logs/mirroring_errors_YYYYMMdd_HHmmss.txt`).

. Verify that mirrored images and catalogs are available using the following the commands:
+
[source,terminal]
----
$ oc get catalogsource -n openshift-marketplace
$ oc get imagedigestmirrorset,imagetagmirrorset -n openshift-config
----
+
For more information, refer to "Mirroring images for a disconnected installation using oc-mirror plugin v2".

. [Optional] To remove images mirrored using oc-mirror plugin v1, follow these steps:
+

> [IMPORTANT]

Reviewer 4: Since you added the admonition to the step below, I think you can delete it here, since you don't need it to be duplicated here

====
oc-mirror plugin v2 does not automatically prune the destination registry, unlike oc-mirror plugin v1. If you want to clean up images no longer needed, use the delete functionality in v2 with `--delete-v1-images`.

Once all images mirrored with oc-mirror plugin v1 are removed, you no longer need to use this flag. If you need to delete images mirrored with oc-mirror plugin v2, do not set `--delete-v1-images`.
====

> .. Mirror the image using oc-mirror plugin v1.

Author: it must be images.

> .. Upgrade the `ImageSetConfiguration` API version from `v1alpha2` (v1) to `v2alpha1` (v2), then run the following command:

Reviewer 4: I am not reviewing for technical accuracy, but from my limited understanding, I wonder if this step is skipping over important info and/or somewhat redundant. 

On one hand, if this is saying to update a field in a file, then it should be more explicit like this:
```suggestion
.. Update the API version in the `ImageSetConfiguration` file from `v1alpha2` (v1) to `v2alpha1` (v2), then run the following command:
```

But on the other hand, is that API update already covered by step 1 of this procedure? If so, then this step should be reword to something like this:
```suggestion
.. [Do some action] by running the following command:
```

Author: 1st scenario.

+
[source,terminal]
----
$ oc-mirror -c isc.yaml file://some-dir --v2
----
+
[NOTE]
====

> `storageConfig` is not a valid field in `ImageSetConfiguration and `DeleteImageSetConfiguration`. Remove this field when updating to oc-mirror plugin v2.

Reviewer 4: ```suggestion
`storageConfig` is not a valid field in the `ImageSetConfiguration` and `DeleteImageSetConfiguration` files. Remove this field when updating to oc-mirror plugin v2.
```

====

> .. Generate a delete manifest and delete v1 images using the following command:

Reviewer 4: Per the [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command):
```suggestion
.. Generate a delete manifest and delete v1 images by running the following command:
```

+
[source,terminal]
----
$ oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://registry.example:5000  --v2
----
+
For more information about deleting images, see "Deletion of images from your disconnected environment".

> .. Delete images based on the generated manifest:

Reviewer 4: ```suggestion
.. Delete images based on the generated manifest by running the following command:
```

// Module included in the following assemblies:
//
// * oc-mirror-migration-v1-to-v2.adoc

:_mod-docs-content-type: PROCEDURE
[id="oc-mirror-migration-process_{context}"]
= Migrating from oc-mirror plugin v1 to oc-mirror plugin v2

> Migrating from oc-mirror v1 to v2 requires manually updating the ImageSetConfiguration, adjusting mirroring commands, and ensuring proper cleanup of v1 artifacts. Follow these steps to transition smoothly:

Reviewer 5: I like this section a lot because it is a straightforward list of steps to do in order to switch from using v1 to using v2.
my personal feeling is that the first module (differences) is a bit redundant... there are only a few information from the first module that are missing in this one, but that we could add here maybe.
Again, this is only my opinion... let's also let the others pitch in here.

      packages:
        - name: aws-load-balancer-operator
----

. Commands have changed to align with oc-mirror plugin v2's new workflow.

.. For mirror-to-disk, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml file://some-dir --v2
----

.. For disk-to-mirror, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml --from file://some-dir docker://remote-registry --v2
----

.. For mirror-to-mirror, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml --from file://some-dir docker://remote-registry --v2
----
+
[NOTE]
====
`--workspace` is now required for mirror-to-mirror operation.
====

. In v2, mirrored content is stored in different locations. You must check the cluster-resources directory inside their working directory for IDMS, ITMS, `CatalogSource`, and `ClusterCatalog` using the following command :
+
[source,terminal]
----
$ ls <workspace>/cluster-resources/
----

. oc-mirror plugin v2 replaces ICSP with IDMS/ITMS. Before applying the new configuration, delete the outdated ICSP entries to avoid conflicts using the following command:
+
[source,terminal]
----

> $ oc delete imagecontentsourcepolicy --all

Reviewer 5: this is rather risky : if ICSP resources other than the ones previously generated by oc-mirror were applied, they would all be removed...
I know this is not so great, but can we talk about removing just the ones generated by oc-mirror.
There is no easy way to identify them unfortunately.
@lmzuccarelli @aguidirh @r4f4 I think it would be a good idea to introduce a label in all the custom resources we generate to allow an admin to easily identify the custom resources generated by oc-mirror in the future. WDYT?

Author: Remove --all and add the label for the icsp name here.

----

> . Remove images mirrored using oc-mirror plugin v1 using the following command:

Reviewer 5: This is more of an optional step.
We could introduce it by saying that v2 doesn' t auto-prune the destination registry, the way v1 does.
So we could tell users who are interrested in pruning images that are no longer needed to refer to the delete functionality of v2, with the reminder to use --delete-v1-images for more compatibility. 
We could say that once all v1 mirrored images have been deleted, the flag is no longer needed.
To be more precise, if they want to delete images mirrored with v2, they shouldn' t be setting that flag.

Reviewer: The workflow to remove images mirrored by oc-mirror v1 should be : 
1) mirror the image by oc-mirror v1 (what error the mirror2disk+disk2mirror / mirror2mirror);
2) The ImageSetConfiguration API version changes from v1alpha2 (v1) to v2alpha1 (v2) , then run the v2 plugin with command : `oc-mirror -c isc.yaml file://anypath --v2`
3) ` oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://the-same-path-as-step2  docker://remote-registry --v2`
4) `oc-mirror delete --delete-yaml-file  the-same-path-as-step2/working-dir/delete/delete-images.yaml docker://remote-registry --v2`

+
[source,terminal]
----
$ oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://remote-registry --v2
----
+
[source,terminal]
----
oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://remote-registry --v2
----
+
[IMPORTANT]
====

> Operator index images are not deleted automatically. You must add them manually to `DeleteImageSetConfiguration`.

Reviewer 5: Let' s tell them also the how and the why:
* maybe instead of saying index images , let' s talk about operator catalog images, which I think will be more identifiable
* v2 doesn' t delete catalog images because, in case the delete workflow was aiming at removing 1 or more operators belonging to that catalog from the registry, the catalog image should not be removed, because it would prevent the cluster, and OLM from functioning correctly
* If the user is sure that no more operators belonging to that catalog are on the registry or running in the cluster, they may force the deletion of the catalog image by explicitly adding it to the list of additionalImages in the deleteImageSetConfiguration

Reviewer: Or say `catalog index image`

====

.Verification

. Once migration is complete, verify that mirrored images and catalogs are available using the following the commands:

> +

Reviewer 5: having the custom resources applied to the cluster doesn't garantee that the images are available unfortunately.
The fact that no errors and no warnings occurred during mirroring, and that no error file was generated (working-dir/logs/mirroring_errors_YYYYMMdd_HHmmss.txt) is a better garantee.

mirror:
  platform:
    channels:
      - name: stable-4.17
    graph: true
  helm:
    repositories:
      - name: sbo
        url: https://redhat-developer.github.io/service-binding-operator-helm-chart/
  additionalImages:
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/openshifttest/hello-openshift@sha256:example_hash
  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
storageConfig:  # REMOVE this field in v2
  local:
    path: /var/lib/oc-mirror
----
+
.Example output with oc-mirror plugin v2 configuration:
[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
      - name: stable-4.17
    graph: true
  helm:
    repositories:
      - name: sbo
        url: https://redhat-developer.github.io/service-binding-operator-helm-chart/
  additionalImages:
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/openshifttest/hello-openshift@sha256:example_hash
  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
----

> . Commands have changed to align with oc-mirror plugin v2's new workflow.

Reviewer: This is not the proper place to introduce the v2 command , better to change to `Command line changes`

  additionalImages:
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/openshifttest/hello-openshift@sha256:example_hash
  operators:
    - catalog: oci:///test/redhat-operator-index
      packages:
        - name: aws-load-balancer-operator
----

. Commands have changed to align with oc-mirror plugin v2's new workflow.

.. For mirror-to-disk, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml file://some-dir --v2
----

.. For disk-to-mirror, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml --from file://some-dir docker://remote-registry --v2
----

.. For mirror-to-mirror, run the following command:
+
[source,terminal]
----
$ oc-mirror --config isc.yaml --workspace file://some-dir docker://remote-registry --v2
----
+
[NOTE]
====
`--workspace` is now required for mirror-to-mirror operation.
====

. In v2, mirrored content is stored in different locations. You must check the cluster-resources directory inside their working directory for IDMS, ITMS, `CatalogSource`, and `ClusterCatalog` using the following command :
+
[source,terminal]
----

> $ ls <workspace>/cluster-resources/

Reviewer 5: I'm not sure this is extra clear for users.
Suggestion: in the previous command line examples, we are using `file://some-dir` .
In the `Mirroring images for a disconnected installation by using the oc-mirror plugin v2`, we are using `file://<file_path>`.
Maybe we should identify this as `file://<v2_workspace` (or any other suggestion: v2_work_path, v2_workspace_path).
This is a way to suggest to the user this is the same folder for m2d, d2m and m2m.
This way, we could say clearly that the yaml files  are in 
```suggestion
$ ls <v2_workspace>/working-dir/cluster-resources/
```

----

> . [*Optional*] oc-mirror plugin v2 replaces ICSP with IDMS/ITMS. Before applying the new configuration, delete the outdated ICSP entries to avoid conflicts using the following command:

Reviewer: as we discussed, better to follow the step to remove the ICSP : delete the installed operator -> delete all v1 images -> delete the cs created by v1 -> delete the icsp

Reviewer 5: we could point them to the section _Configuring your cluster to use the resources generated by oc-mirror plugin v2)_ for exact instructions.

As discussed and mentioned by @zhouying7780 , we only want them to delete icsp in case all the corresponding v1 images were deleted

Author: Remove the step. Add a note redirecting the users to v2 docs.

+
[source,terminal]
----
$ oc delete <ICSP/IDMS/ITMS> <1>
----
<1> Specify the `ICSP/IDMS/ITMS` name.

. To remove images mirrored using oc-mirror plugin v1, follow these steps:
+
[IMPORTANT]
====
oc-mirror plugin v2 does not automatically prune the destination registry, unlike oc-mirror plugin v1. If you want to clean up images no longer needed, use the delete functionality in v2 with `--delete-v1-images`.

Once all images mirrored with oc-mirror plugin v1 are removed, you no longer need to use this flag. If you need to delete images mirrored with oc-mirror plugin v2, do not set `--delete-v1-images`.
====

.. Mirror the image using oc-mirror plugin v1.

.. Upgrade the `ImageSetConfiguration` API version from `v1alpha2` (v1) to `v2alpha1` (v2), then run the following command:
+
[source,terminal]
----
$ oc-mirror -c isc.yaml file://some-dir --v2

> ----

Reviewer 5: @zhouying7780, @lmzuccarelli,
is it mandatory that user runs a v2 mirror to disk prior to starting the delete workflow for v1 images?
Apologies for my bad memory on this subject

Reviewer 3: @sherine-k I think so, the delete uses the same collector of the mirror (copy) so it will check the cache in order to collect the images (it always runs on top of diskToMirror). 

If there is nothing in the cache, so it won't be able to collect the images to be deleted.

Reviewer 5: thanks!
This makes the delete workflow bit long then... Food for thought...
I'm wondering if a dry-run would slightly help...

.. Generate a delete manifest and delete v1 images using the following command:
+
[source,terminal]
----

> $ oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://remote-registry --v2

Reviewer: Better to emphasize same `remote-registry` path as the v1 mirror used.

Author: Using `docker://registry.example:5000 ` as specified in v1.

> $ oc-mirror delete --config=delete-isc.yaml --generate --delete-v1-images --workspace file://some-dir docker://remote-registry --v2

Reviewer 6: @subhtk - we introduce the concept of --delete-isc.yaml - maybe we should give them an example of one to use i.e 

- change the ImageSetconfiguration -> DeleteImageSetConfiguration.
- change mirror to delete
- save as delete-isc.yaml

Author: We can add an example, I asked another writer as well, but duplication of the existing content should be avoided.

----

.. Delete images based on the generated manifest:
+
[source,terminal]
----
$ oc-mirror delete --delete-yaml-file some-dir/working-dir/delete/delete-images.yaml docker://remote-registry --v2
----

.Verification

> . Once the migration is complete, verify that mirrored images and catalogs are available. Ensure:

Reviewer 6: @subhtk - WOuld it make sense to point out that garbage collection varies from registry to registry and in some cases may need the sys-admin of the registry to perform a manual garbage collect  to ensure all relevant images are actually deleted

Reviewer 5: a very interesting point.
I think it absolutely needs to  be mentioned in `modules/oc-mirror-workflows-delete-v2.adoc` at least

- No errors or warnings occurred during mirroring.
- No error file was generated (`working-dir/logs/mirroring_errors_YYYYMMdd_HHmmss.txt`).

. Verify that mirrored images and catalogs are available using the following the commands:
+
[source,terminal]
----
$ oc get catalogsource -n openshift-marketplace
$ oc get imagedigestmirrorset,imagetagmirrorset -n openshift-config
----

> For more information, refer to _mirroring in a disconnected environment_.

Reviewer 5: we don't tell them to apply these to the cluster first...
so if they follow this migration guide to the letter, catalogsource hasn' t been applied yet...
I would rather we point them back to the v2 documentation (Configuring your cluster to use the resources generated by oc-mirror plugin v2) instead
